### PR TITLE
Correctly deserialize `undefined` unstable_cache data

### PR DESCRIPTION
**Pull request by mattddean**
The value `undefined` can be saved to the incremental cache as `undefined` (`JSON.stringify(undefined)`) with no errors, but when retrieving it, we attempt to parse it as JSON using `JSON.parse(undefined)`. This throws an error. We should instead deserialize `undefined` as `undefined` when retrieving.

relevant discussion: #59087

### CODE REVIEW

**Review by prasad-pul — APPROVED** [review]
Looks good!

File: packages/next/src/server/web/spec-extension/unstable-cache.ts [hunk 1]
```diff
@@ -158,7 +158,9 @@ export function unstable_cache<T extends Callback>(
 
         if (cacheEntry) {
           const resData = cacheEntry.value.data
-          cachedValue = JSON.parse(resData.body)
+          // JSON.stringify(undefined) returns `undefined`, but JSON.parse(undefined) throws an error
+          cachedValue =
+            resData.body !== undefined ? JSON.parse(resData.body) : undefined
         }
```
Comment (smaeda-ks) [inline packages/next/src/server/web/spec-extension/unstable-cache.ts:164]: Can we also just move this entire condition below `if (isStale) { ... }` ?

Comment (mattddean) [reply packages/next/src/server/web/spec-extension/unstable-cache.ts]: makes sense, updated

Comment (mattddean) [reply packages/next/src/server/web/spec-extension/unstable-cache.ts]: The function was refactored, but the issue of [`JSON.stringify`ing the cache's data](https://github.com/vercel/next.js/blob/c754d9fb6ecddc7cf46a61548dc8c49539c398d1/packages/next/src/server/web/spec-extension/unstable-cache.ts#L35) that could be `undefined`, then [`JSON.parse`ing it](https://github.com/vercel/next.js/blob/c754d9fb6ecddc7cf46a61548dc8c49539c398d1/packages/next/src/server/web/spec-extension/unstable-cache.ts#L186) on retrieval persists, so I'm going to leave this PR open.

Comment (prasad-pul) [reply packages/next/src/server/web/spec-extension/unstable-cache.ts]: I'm not sure why is this not being merged or reviewed yet by NextJs team 🤔
Isn't this a critical bug, when an api returns `undefined` the `unstable_cache` crashes the Application. Isn't that a very common scenario??

cc: @mattddean , @ijjk , @timneutkens

**Comment by stabildev** [comment]
important!

**Comment by prasad-pul** [comment]
No idea why this isn't looked up by Next js team??

**Comment by samcx** [comment]
Taking a look!

**Comment by samcx** [comment]
Still looking into this everyone 🙇🏼

**Comment by mattddean** [comment]
> Still looking into this everyone 🙇🏼 

Happy to help if I can! Or to write a failing test if you point me at the right directory to put it in.

**Review by gnoff — APPROVED** [review]

**Comment by samcx** [comment]
@mattddean Yeah we should add a test. Let me see where we should

**Comment by samcx** [comment]
@mattddean You can add a directory called `unstable_cache` (similar to the `no_store` directory, but without the `unstable_noStore`) and add a 1-2 test(s) (one to test `unstable_cache` is working as expected, and another to test undefined`) after the `unstable_noStore` describe → https://github.com/vercel/next.js/blob/a767aa87134274af5061bf1664f7349d1d76f882/test/e2e/app-dir/app-static/app-static.test.ts#L2991.

Let me know if that helps!

**Comment by mattddean** [comment]
@samcx Super helpful, thanks! Let me know if the tests I've added look sufficient to you.

**Comment by mattddean** [comment]
Quick screenshot of the new test failing before applying this change:

Screenshot 2024-03-08 at 5 44 28 PM

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 7b5e8f0d7bf9d9048a0cb8dfe75cfa10b639b3c9

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 7b5e8f0d7bf9d9048a0cb8dfe75cfa10b639b3c9

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 7b5e8f0d7bf9d9048a0cb8dfe75cfa10b639b3c9

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.7s | N/A |
| buildDurationCached | 7.4s | 5.8s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +924 B |
| nextStartRea..uration (ms) | 429ms | 435ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 529 B | 529 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.06 kB | 3.07 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.5 kB | 96.5 kB | ✓ |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 92.7 kB | 92.7 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.2 kB | 91.2 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | mattddean/next.js patch-2 | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 104 kB | 105 kB | ⚠️ +888 B |
| Overall change | 104 kB | 105 kB | ⚠️ +888 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 8ad3d50dceaaa51009fcb64dc6a7f19a5138bed2

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 8ad3d50dceaaa51009fcb64dc6a7f19a5138bed2

`pnpm test-start test/e2e/app-dir/app-static/app-static-custom-handler.test.ts`
- app-dir static/dynamic handling > should output HTML/RSC files for static paths

<details>
<summary>Expand output</summary>

  ● app-dir static/dynamic handling › should output HTML/RSC files for static paths

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `app-dir static/dynamic handling should output HTML/RSC files for static paths 1`

    - Snapshot  - 0
    + Received  + 4

    @@ -168,10 +168,14 @@
        "stale-cache-serving/route-handler/route.js",
        "static-to-dynamic-error-forced/[id]/page.js",
        "static-to-dynamic-error-forced/[id]/page_client-reference-manifest.js",
        "static-to-dynamic-error/[id]/page.js",
        "static-to-dynamic-error/[id]/page_client-reference-manifest.js",
    +   "unstable-cache/dynamic-undefined/page.js",
    +   "unstable-cache/dynamic-undefined/page_client-reference-manifest.js",
    +   "unstable-cache/dynamic/page.js",
    +   "unstable-cache/dynamic/page_client-reference-manifest.js",
        "variable-config-revalidate/revalidate-3.html",
        "variable-config-revalidate/revalidate-3.rsc",
        "variable-config-revalidate/revalidate-3/page.js",
        "variable-config-revalidate/revalidate-3/page_client-reference-manifest.js",
        "variable-revalidate-edge/body/page.js",

      495 |           })
      496 |
    > 497 |         expect(files.sort()).toMatchInlineSnapshot(`
          |                              ^
      498 |           [
      499 |             "(new)/custom/page.js",
      500 |             "(new)/custom/page_client-reference-manifest.js",

      at Object.toMatchInlineSnapshot (e2e/app-dir/app-static/app-static.test.ts:497:30)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by Facundo-Martin** [comment]
Haha just ran into this bug today at work. It's nice to see it's been fixed! Although just for the canary version I guess.
My solution was to just return `null` instead of `undefined`.

**Pull request merged**